File: 跳跃游戏(数组，贪心算法)/跳跃游戏.java
//维护一个可以到达的最大距离的数组，如果下一位的坐标大于他这返回false
public class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int rightmost = 0;
        for (int i = 0; i < n; ++i) {
            if (i <= rightmost) {
                rightmost = Math.max(rightmost, i + nums[i]);
                if (rightmost >= n - 1) {
                    return true;
                }
            }
        }
        return false;
    }
}
//计算最短跳跃距离
//从后往前选择距离最远的点，从头开始遍历
class Solution {
    public int jump(int[] nums) {
        int position = nums.length - 1;
        int steps = 0;
        while (position > 0) {
            for (int i = 0; i < position; i++) {
                if (i + nums[i] >= position) {
                    position = i;
                    steps++;
                    break;
                }
            }
        }
        return steps;
    }
}
//从前往后寻找跳跃距离最远的点
//具体的实现中，我们维护当前能够到达的最大下标位置，记为边界。我们从左到右遍历数组，到达边界时，更新边界并将跳跃次数增加 1
class Solution {
    public int jump(int[] nums) {
        int length = nums.length;
        int end = 0;
        int maxPosition = 0; 
        int steps = 0;
        for (int i = 0; i < length - 1; i++) {
            maxPosition = Math.max(maxPosition, i + nums[i]); 
            if (i == end) {
                end = maxPosition;
                steps++;
            }
        }
        return steps;
    }
}

File: 跳跃游戏(数组，贪心算法)/跳跃游戏.cpp
//第一次自己的想法，但会超出时间限制
bool jump(vector<int>& nums,int start){
        int i;
        if(start == nums.size()-1)
            return true;
        for(i = 1;i<=nums[start];i++){
            if(jump(nums,i+start))
                return true;
        }
        return false;
    }
    bool canJump(vector<int>& nums) {
        return jump(nums,0);
}
//改进后的写法
//1.如果某一个作为 起跳点 的格子可以跳跃的距离是 3，那么表示后面 3 个格子都可以作为 起跳点。
//2.可以对每一个能作为 起跳点 的格子都尝试跳一次，把 能跳到最远的距离 不断更新。
//3.如果可以一直跳到最后，就成功了。
bool canJump(vector<int>& nums) 
{
	int k = 0;
	for (int i = 0; i < nums.size(); i++)
	{
		if (i > k) return false;
		k = max(k, i + nums[i]);
	}
	return true;
}
